File: docs-source/docs/modules/how-to/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCart.scala
package shopping.cart

import java.time.Instant

import org.apache.pekko.actor.typed.ActorSystem
import org.apache.pekko.actor.typed.Behavior
import org.apache.pekko.cluster.sharding.typed.scaladsl.ClusterSharding
import org.apache.pekko.cluster.sharding.typed.scaladsl.Entity
import org.apache.pekko.cluster.sharding.typed.scaladsl.EntityContext
import org.apache.pekko.cluster.sharding.typed.scaladsl.EntityTypeKey

object ShoppingCart {

  // tag::howto-crud-to-es-initial[]
  sealed trait Command
  final case class AddItem(itemId: String, quantity: Int) extends Command
  case object Checkout extends Command

  sealed trait Event {
    def cartId: String
  }
  final case class ItemAdded(cartId: String, itemId: String, quantity: Int)
      extends Event
  final case class CheckedOut(cartId: String, eventTime: Instant) extends Event
  // end::howto-crud-to-es-initial[]

  val EntityKey: EntityTypeKey[Command] = ???
  def apply(cartId: String, projectionTag: String): Behavior[Command] = ???
  val tags: Seq[String] = ???

  // tag::write-side-with-role[]
  def init(system: ActorSystem[_]): Unit = {
    val behaviorFactory: EntityContext[Command] => Behavior[Command] = {
      entityContext =>
        val i = math.abs(entityContext.entityId.hashCode % tags.size)
        val selectedTag = tags(i)
        ShoppingCart(entityContext.entityId, selectedTag)
    }
    ClusterSharding(system).init(
      Entity(EntityKey)(behaviorFactory).withRole("entity")
    ) // <1>
  }
  // end::write-side-with-role[]

}

File: docs-source/docs/modules/how-to/examples/shopping-cart-service-java/src/main/java/shopping/cart/ShoppingCart.java
package shopping.cart;

import java.time.Duration;
import java.time.Instant;
import java.util.*;
import org.apache.pekko.actor.typed.ActorRef;
import org.apache.pekko.actor.typed.ActorSystem;
import org.apache.pekko.actor.typed.Behavior;
import org.apache.pekko.actor.typed.SupervisorStrategy;
import org.apache.pekko.actor.typed.javadsl.Behaviors;
import org.apache.pekko.cluster.sharding.typed.javadsl.ClusterSharding;
import org.apache.pekko.cluster.sharding.typed.javadsl.Entity;
import org.apache.pekko.cluster.sharding.typed.javadsl.EntityTypeKey;
import org.apache.pekko.pattern.StatusReply;
import org.apache.pekko.persistence.typed.PersistenceId;
import org.apache.pekko.persistence.typed.javadsl.*;

public final class ShoppingCart
    extends EventSourcedBehaviorWithEnforcedReplies<
        ShoppingCart.Command, ShoppingCart.Event, ShoppingCart.State> {

  static final class State {}

  interface Command {}

  public static final class AddItem implements Command {
    final String itemId;
    final int quantity;
    final ActorRef<StatusReply<Summary>> replyTo;

    public AddItem(String itemId, int quantity, ActorRef<StatusReply<Summary>> replyTo) {
      this.itemId = itemId;
      this.quantity = quantity;
      this.replyTo = replyTo;
    }
  }

  public static final class Checkout implements Command {
    final ActorRef<StatusReply<Summary>> replyTo;

    public Checkout(ActorRef<StatusReply<Summary>> replyTo) {
      this.replyTo = replyTo;
    }
  }

  public static final class Summary {
    final Map<String, Integer> items;
    final boolean checkedOut;

    public Summary(Map<String, Integer> items, boolean checkedOut) {
      // defensive copy since items is a mutable object
      this.items = new HashMap<>(items);
      this.checkedOut = checkedOut;
    }
  }

  abstract static class Event {
    public final String cartId;

    public Event(String cartId) {
      this.cartId = cartId;
    }
  }

  static final class ItemAdded extends Event {
    public final String itemId;
    public final int quantity;

    public ItemAdded(String cartId, String itemId, int quantity) {
      super(cartId);
      this.itemId = itemId;
      this.quantity = quantity;
    }
  }

  static final class CheckedOut extends Event {
    final Instant eventTime;

    public CheckedOut(String cartId, Instant eventTime) {
      super(cartId);
      this.eventTime = eventTime;
    }
  }

  static final EntityTypeKey<Command> ENTITY_KEY =
      EntityTypeKey.create(Command.class, "ShoppingCart");

  static final List<String> TAGS =
      Collections.unmodifiableList(
          Arrays.asList("carts-0", "carts-1", "carts-2", "carts-3", "carts-4"));

  // tag::write-side-with-role[]
  public static void init(ActorSystem<?> system) {
    ClusterSharding.get(system)
        .init(
            Entity.of(
                ENTITY_KEY,
                entityContext -> {
                  int i = Math.abs(entityContext.getEntityId().hashCode() % TAGS.size());
                  String selectedTag = TAGS.get(i);
                  return ShoppingCart.create(entityContext.getEntityId(), selectedTag);
                }));
  }
  // end::write-side-with-role[]

  public static Behavior<Command> create(String cartId, String projectionTag) {
    return Behaviors.setup(
        ctx -> EventSourcedBehavior.start(new ShoppingCart(cartId, projectionTag), ctx));
  }

  private final String projectionTag;

  private final String cartId;

  private ShoppingCart(String cartId, String projectionTag) {
    super(
        PersistenceId.of(ENTITY_KEY.name(), cartId),
        SupervisorStrategy.restartWithBackoff(Duration.ofMillis(200), Duration.ofSeconds(5), 0.1));
    this.cartId = cartId;
    this.projectionTag = projectionTag;
  }

  @Override
  public Set<String> tagsFor(Event event) {
    return Collections.singleton(projectionTag);
  }

  @Override
  public State emptyState() {
    return new State();
  }

  @Override
  public CommandHandlerWithReply<Command, Event, State> commandHandler() {
    return null; // not implemented here
  }

  @Override
  public EventHandler<State, Event> eventHandler() {
    return null; // not implemented here
  }
}
